### PR TITLE
New: sharing state core concept!

### PR DESCRIPTION
**Pull request by bholmesdev**
#### What kind of changes does this PR include?

- [ ] Minor content fixes (broken links, typos, etc.)
- [X] New or updated content
- [ ] Translated content
- [X] Changes to the docs site code
- Something else!

#### Description

- Add new `Tabs` Preact component for tab views across our docs. This includes:
  - A `UIFrameworkTabs` utility component for easily embedding UI framework comparisons
  - A generic `Tabs` component for any tab view you could want, based on slots
- Add a new core concept page: Sharing State. This includes:
  - A primer on what user problem this page addresses
  - Why nanostores (with a brief comparison to Svelte stores)
  - A usage example that _will_ mirror one of our astro.new examples soon
  - A walkthrough on nanostore "atoms"
  - A walkthrough on nanostore "maps"
- Add `LoopingVideo.astro` utility for video embeds

https://user-images.githubusercontent.com/51384119/177429138-e123dcd0-3313-49a0-b97d-def21aab5f65.mov

### CODE REVIEW

File: src/pages/en/core-concepts/sharing-state.md
Comment (Yan-Thomas) [inline src/pages/en/core-concepts/sharing-state.md:37]:
```suggestion
If you want to avoid third-party libraries, [Svelte stores](https://svelte.dev/tutorial/writable-stores) are a great cross-island communication tool on their own. Still, you might prefer nanostores if a) you like their add-ons for ["objects"](https://github.com/nanostores/nanostores#maps) and [async state](https://github.com/nanostores/nanostores#lazy-stores), or b) you want to communicate between Svelte and other UI frameworks like Preact or Vue.
```

File: src/pages/en/core-concepts/sharing-state.md
Comment (Yan-Thomas) [inline src/pages/en/core-concepts/sharing-state.md:77]:
```suggestion
You can jump into the [nanostores usage guide](https://github.com/nanostores/nanostores#guide) from here, or follow along with our example below!
```

File: src/i18n/en/nav.ts
Comment (Yan-Thomas) [inline src/i18n/en/nav.ts:23]:
```suggestion
	{ text: 'Sharing State', slug: 'core-concepts/sharing-state', key: 'core-concepts/sharing-state' },
```

**Comment by crutchcorn** [comment]
Awesome work! 🎉 

Something I'm noticing is the "Jump when tab switch" due to a mismatch of content length between tabs. This leads to a jarring experience that, when I implemented it for my blog, lead to a lot of errant bug reports being filed.

Here's a GIF of this in an extreme example:

https://user-images.githubusercontent.com/9100169/174496071-0448ee34-f992-4bd8-9e0c-c55d7ec0d8b5.mp4

To fix this, I effectively:

- Attached a unique ID to each tab header
- On tab selection change, added a `setTimeout(() => {}, 0)` to queue a microtask to jump to the tab header using `scrollIntoView`

Here's a PR that I wrote to fix this problem:

https://github.com/unicorn-utterances/unicorn-utterances/pull/374

File: src/components/tabs/Tabs.module.css
Comment (Yan-Thomas) [inline src/components/tabs/Tabs.module.css:13]: The other tabs are using `--theme-divider` for the color and 4px as its size, I think it would be better if we used the same values for all the Tabs. Also, maybe we should expand the divider line to fill the whole screen as the Installation Tab does?

I'd also add `cursor: pointer` just so it's clear we can click the tab

```suggestion
	border-bottom: 4px solid var(--theme-divider);
	cursor: pointer;
```

File: src/pages/en/core-concepts/sharing-state.md
Comment (Yan-Thomas) [inline src/pages/en/core-concepts/sharing-state.md:387]: I'm not sure about using `<details />` inside notes, I think we could just use a custom title instead:

```suggestion
:::note [**🙋 Why can use use `.get()` here instead of a `useStore` helper?**]
You may have noticed we're calling `cartItems.get()` here, instead of grabbing that `useStore` helper from our React / Preact / Solid / Vue examples. This is because **useStore is meant to trigger component re-renders.** In other words, `useStore` should be used whenever the store value is being rendered to the UI. Since and we're reading the value when an **event** is triggered (`addToCart` in this case), and we aren't trying to render that value, we don't need `useStore` here.
:::
```

**Review by Yan-Thomas — COMMENTED** [review]
Amazing work Ben "Tabs" H! I left a few suggestions for you 🙌

I also verified the Tabs component accessibility with NVDA and it is perfect for me 👌

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:387]: Hm, I kinda prefer putting questions inside `details` elements so that, if the reader _didn't_ have that question, they can happily skip over it. Custom `note` headers are news to me though! I'd be happy to change this one if you feel strongly enough, but I definitely want to keep the Svelte stores callout inside a details block.

File: src/components/tabs/Tabs.module.css
Comment (bholmesdev) [reply src/components/tabs/Tabs.module.css:13]: Sure, we could extend the divider! Agreed on the suggestion too 👍

**Comment by zadeviggers** [comment]
This looks great! I did notice that on mobile (android firefox) some of the tabs overflow off the page and you can't scroll them to see the cut off content.
I also think it might be good to touch on the Solid stores a bit more in depth since the page is about sharing state in general, not nanostores specificaly (unless I'm very much mistaken, in which case I apologise).
Overall this is a very well written page!

File: src/pages/en/core-concepts/sharing-state.md
Comment (natemoo-re) [inline src/pages/en/core-concepts/sharing-state.md:30]: This is fine to do as a follow-up, but I feel like we should have a `:::faq` component that has `details` and `summary` built-in? These don't semantically feel like a note.

File: src/pages/en/core-concepts/sharing-state.md
Comment (natemoo-re) [inline src/pages/en/core-concepts/sharing-state.md:86]: Would be nice to link to a live StackBlitz example here? Some folks prefer to dive right into a live project.

**Review by natemoo-re — APPROVED** [review]
Looking great! Two small suggestions, neither blocking.

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:30]: Agreed! I'll hold off since this PR adds a lot of tooling already, but that definitely came to my mind.

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:86]: Updating our `with-nanostores` template to use this very example! I'll embed a link once that's merged

Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:86]: Updated!

**Comment by bholmesdev** [comment]
@zadeviggers Also added a callout on solid stores for shared state. Let me know if these instructions make sense!
https://github.com/withastro/docs/pull/934/commits/e8123d1c255ee9161bf7b79764a77185d3721a59

**Comment by bholmesdev** [comment]
@crutchcorn addressed!

**Comment by zadeviggers** [comment]
Looks great @bholmesdev!

**Comment by sarah11918** [comment]
Just a quick note before I check the sites I have queued up to read re: "using emoji as end punctuation".... BEN...

I'm not sure this page fits in the "Core Concepts" section, and I'm leaning towards it belonging in "Features" rather than "Basics" (as the pages are structured now). I'm curious to hear others' thoughts on this.

The material that Fred is drafting in #893 is more overall description/philosophy of principle, and not technical walkthroughs. "Basics" are (right now) for things that are inherent to understanding an Astro project: file structure, components, pages/routing etc. and "Features" (formerly "Guides") are "Things you'd expect a website to have/do, and here's how you'd do that in/with your Astro site." The Basics briefly show some essential code patterns, but the Features will demonstrate the code required to achieve ... a feature. This feels like the latter to me?  Thoughts?

File: src/pages/en/core-concepts/sharing-state.md
Comment (Yan-Thomas) [inline src/pages/en/core-concepts/sharing-state.md:405]: Just found these typos
```suggestion
<summary>**🙋 Why use `.get()` here instead of a `useStore` helper?**</summary>

You may have noticed we're calling `cartItems.get()` here, instead of grabbing that `useStore` helper from our React / Preact / Solid / Vue examples. This is because **useStore is meant to trigger component re-renders.** In other words, `useStore` should be used whenever the store value is being rendered to the UI. Since we're reading the value when an **event** is triggered (`addToCart` in this case), and we aren't trying to render that value, we don't need `useStore` here.
</details>
```

**Comment by hippotastic** [comment]
I love the general idea of this! However, switching any of the tabs in the Netlify preview doesn't work on iOS Safari for me.

**Comment by hippotastic** [comment]
It works well in iOS Safari now after your updates. Great job!

**Comment by FredKSchott** [comment]
Like @sarah11918 said, this is more of a Guide than a technical concept essential to the very core of Astro :) Place it with the other features in the sidebar, and we can experiment with renaming that heading to "Guides" if it makes more sense.

File: src/components/tabs/Tabs.tsx
Comment (delucis) [inline src/components/tabs/Tabs.tsx:56]: There’s no way to sensibly scope this to _only_ Safari right? I’ve been playing around to try to avoid the scroll up, but can’t figure out a nicer behaviour. Without this works perfectly for me in FF & Chrome, so in a way this actually _worsens_ the UX there, but that may be unavoidable in our quest to make Safari not awful.

File: src/components/tabs/Tabs.module.css
Comment (delucis) [inline src/components/tabs/Tabs.module.css:28]: Looks like focus styling is getting clipped now that the tab list can be scrolled:

image


How about something like this to fix that?

```suggestion

.tab:focus {
	outline: 2px solid var(--theme-accent);
	outline-offset: -2px;
}

```

image

File: src/components/tabs/Tabs.tsx
Comment (delucis) [inline src/components/tabs/Tabs.tsx:92]: We’re ending up with multiple identical `id`s on this page, which isn’t valid: https://dequeuniversity.com/rules/axe/4.4/duplicate-id-aria

I swear I saw you mention using incrementing IDs somewhere, maybe on Discord? Did that not work out?

File: src/pages/en/core-concepts/sharing-state.md
Comment (delucis) [inline src/pages/en/core-concepts/sharing-state.md:313]: Do we have a way to style the tab labels here? Elsewhere in docs we would use backticks to wrap filenames in `<code>` and it might look better here too?

Another option would be to label these “JavaScript” and “TypeScript” and include the filename in the code sample? A standard way to associate filenames with code samples is on my stack in #918 so this could be about to be standardised in any case, so the more generic tab labelling may make more sense in any case once we have that.

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:12]: I'm very sad to see this happen, but I had to try...

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:15]: I'll remove it before I get hit with requested changes 😉

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:43]: Didn't know custom titles were possible! Love this 👏

File: src/pages/en/core-concepts/sharing-state.md
Comment (sarah11918) [inline src/pages/en/core-concepts/sharing-state.md:300]: "we may store your cart" sounds a little unclear: you *might*? You *are allowed to*? 

I think "will" (if certain) or "can" (if a possible option that you're choosing to employ) both sound a bit more standard.

File: src/pages/en/core-concepts/sharing-state.md
Comment (sarah11918) [inline src/pages/en/core-concepts/sharing-state.md:349]:
```suggestion
- **If that item doesn't exist in your cart**, add the item with a starting quantity of 1.
- **If that item _does_ already exist**, bump the quantity by 1.
```
nit: bold the words, not the punctuation mark is our MO

File: src/pages/en/core-concepts/sharing-state.md
Comment (sarah11918) [inline src/pages/en/core-concepts/sharing-state.md:681]: AP style guide says "e-commerce." Shopify, WordPress and Amazon use ecommerce. ... in our Astro vs X page we say eCommerce...  🤪 

I'm OK with ecommerce, and will make a note to update elsewhere. Just wanted it to be known that I checked, for the record!

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md:57]: Inherited from auto and manual tabs 👏

**Review by sarah11918 — COMMENTED** [review]
Whew! Ben! Sorry, this took a little long because besides just a proofreading edit, one thing I look closely for is internal consistency both within the page and within the Docs as a whole. And, this, while delightful, takes both a slightly different tone, and a different level of technical explanation than most other pages in the docs. So, it did also require a bit of double-checking for "comps" around the site... all while reading the comments in Discord, seeing the file update underneath me... 

So, here are my thoughts! Thanks for being patient! And thanks as always for contributing to the docs.. and uniting so many people into one PR. It's great to see so many people get involved!

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md]: Updated to use Nano Stores wherever we reference the library name!

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md]: "will be seamless" now!

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md]: using "can!"

File: src/pages/en/core-concepts/sharing-state.md
Comment (bholmesdev) [reply src/pages/en/core-concepts/sharing-state.md]: Updated to the clearer: "Astro recommends a different solution for shared client-side storage"

**Comment by sarah11918** [comment]
Thanks for fi-NIT-shing this off, Ben! Friday merge, FTW!  🥳

**Comment by inwardmovement** [comment]
Were [Alpine stores](https://alpinejs.dev/globals/alpine-store) considered? May we add a comparison with them?

**Pull request merged**
